# Στοχαστικές Διεργασίες και Βελτιστοποίηση στη Μηχανική Μάθηση

## 1ο Εργαστήριο - Logistic Regression

- Ονομ/νυμο: Χρήστος Νίκου
- AM: 03400146
- Ιδιότητα: Μεταπτυχιακός φοιτητής Επιστήμης Δεδομένων και Μηχανικής Μάθησης (ΕΔΕΜΜ)
- Ηλεκτρονική Διεύθυνση: christosnikou@mail.ntua.gr / chrisnick92@gmail.com

<h1><b>Ο αλγόριθμος Logistic Regression</b></h1>

# Εκφώνηση 

<p align="justify">Στην άσκηση αυτή θα μελετήσετε τον αλγόριθμο <i>logistic regression</i>, κατασκευάζοντας ένα σύντομο πρόγραμμα. Στην άσκηση αυτή θα χρησιμοποιήσετε τη βιβλιοθήκη της <i>Python Scikit-Learn</i>. Για τη διευκόλυνσή σας παρέχονται οι δηλώσεις των βιβλιοθηκών που θα χρησιμοποιήσετε καθώς και εντολές με κενά. Συγκεκριμένα, θα κατασκευάσετε έναν <i>ταξινομητή Spam μηνυμάτων SMS</i>. Τα δεδομένα που θα χρησιμοποιήσετε για την εκπαίδευση και την επικύρωση του μοντέλου, καθώς και πληροφορίες για αυτά μπορούν να βρεθούν <a href="https://archive.ics.uci.edu/ml/datasets/sms+spam+collection"><i>εδώ</i></a>.</p>
<p align="justify">Καλείστε να αναπτύξετε ένα πρόγραμμα, το οποίο:</p>
<ul>
<li>Θα φορτώνει τα δεδομένα από το αρχείο <i>.csv</i>.</li>
<li>Θα διαχωρίζει με τυχαίο τρόπο τα δεδομένα που παρέχονται σε δεδομένα εκπαίδευσης <i>(training set)</i> και δεδομένα για τον υπολογισμό της ακρίβειας του μοντέλου <i>(test set)</i>.</li>
<li>Θα πραγματοποιεί προεπεξεργασία στα δεδομένα χρησιμοποιώντας τη μέθοδο <i>TfidVectorizer</i> της βιβλιοθήκης <i>Scikit-Learn</i>. Περισσότερες πληροφορίες για τη μέθοδο <i>TfidVectorizer</i>, που περιλαμβάνεται στις δηλώσεις του προγράμματος, μπορούν να βρεθούν <a href="https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html"><i>εδώ</i></a>. Να σημειωθεί πως η μέθοδος <i>TfidVectorizer</i> της <i>Scikit-Learn</i> αφαιρεί και τα σημεία στίξης, καθώς επεξεργάζεται το dataset.</li>
<li>Θα εκπαιδεύει το μοντέλο <i>logistic regression</i>.
<li>Θα υπολογίζει την ακρίβειά του πάνω στο <i>test set</i>.</li>
</ul>
<p align="justify">Για ποιους λόγους πιστεύετε ότι δεν είναι κατάλληλη η εφαρμογή της μεθόδου <i>linear regression</i> στο συγκεκριμένο πρόβλημα;</p>

# Απαντήσεις

## Ερώτημα 1

> *Φορτώστε τα δεδομένα από το αρχείο .csv από το παρακάτω [link](#https://archive.ics.uci.edu/ml/datasets/sms+spam+collection).*

Παρακάτω κατεβάζουμε το αρχείο .csv και το διαβάζουμε με τη βιβλιοθήκη *pandas*. Παρακάτω βλέπουμε το DataFrame των δεδομένων, επίσης, μετατρέπουμε τα δεδομένα σε *numpy arrays*.

In [1]:
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split, cross_val_score

In [2]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00228/smsspamcollection.zip

from zipfile import ZipFile

file_name = 'smsspamcollection.zip'
with ZipFile(file_name, 'r') as zip:
  zip.extract('SMSSpamCollection')

import pandas as pd

path = 'SMSSpamCollection'
df = pd.read_csv(path, sep='\t', header = None)
features, labels = df.iloc[:,1].to_numpy(), df.iloc[:,0].to_numpy()

--2022-03-15 10:47:26--  https://archive.ics.uci.edu/ml/machine-learning-databases/00228/smsspamcollection.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 203415 (199K) [application/x-httpd-php]
Saving to: 'smsspamcollection.zip.6'

     0K .......... .......... .......... .......... .......... 25% 20.7K 7s
    50K .......... .......... .......... .......... .......... 50%  128K 3s
   100K .......... .......... .......... .......... .......... 75% 71.0K 1s
   150K .......... .......... .......... .......... ........  100% 65.0K=4.3s

2022-03-15 10:47:32 (46.7 KB/s) - 'smsspamcollection.zip.6' saved [203415/203415]



In [3]:
df

0                                                  1
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...
...    ...                                                ...
5567  spam  This is the 2nd time we have tried 2 contact u...
5568   ham               Will ü b going to esplanade fr home?
5569   ham  Pity, * was in mood for that. So...any other s...
5570   ham  The guy did some bitching but I acted like i'd...
5571   ham                         Rofl. Its true to its name

[5572 rows x 2 columns]

Όπως βλέπουμε το σύνολο δεδομένων αποτελείται από 5572 γραμμές και 2 στήλες. Η μια στήλη του συνόλου δεδομένων αντιστοιχεί στα σε γραπτά μηνύματα SMS και η δεύτερη στήλη του συνόλου δεδομένων φέρει την ένδειξη αν το το εν λόγω μήνυμα είναι SPAM ή HAM. Σκοπός μας είναι να αναπτύξουμε έναν ταξινομητή λογιστικής παλινδρόμησης που να κατηγοροποιεί τα μηνύματα σε SPAM και HAM. Το πρώτο που πρέπει να κάνουμε πριν αναπτύξουμε τον ταξινομητή είναι να φέρουμε το σύνολο δεδομένων σε κατάλληλη μορφή ώστε να είναι συμβατή με τις βιβλιοθήκες του *Scikit - learn*.

In [4]:
from sklearn import preprocessing

le = preprocessing.LabelEncoder()
le.fit(labels)
y = le.transform(labels)
X = features

Όπως βλέπουμε παρακάτω μέσω της συνάρτησης *LabelEncoder* έχουμε αντιστοιχίσει τα *labels (ham & spam)* στις τιμές 0 και 1.

In [5]:
print(f"- The true labels are: {y}")
print(f"- And the transformed labels are: {labels}")

- The true labels are: [0 0 1 ... 0 0 0]
- And the transformed labels are: ['ham' 'ham' 'spam' ... 'ham' 'ham' 'ham']


## Ερώτημα 2

> *Διαχωρίστε με τυχαίο τρόπο τα δεδομένα που παρέχονται σε δεδομένα εκπαίδευσης (training set) και δεδομένα για τον υπολογισμό της ακρίβειας του μοντέλου (test set).*

Σε αυτό το σημείο χρησιμοποιώντας τη συνάρτηση *train_test_split* της βιβλιοθήκης *scikit-learn* διαχωρίζουμε το σύνολο των δεδομένων σε σύνολο εκπαίδευσης (*train set*) και στο σύνολο ακρίβειας (*test set*).

In [6]:
X_train_raw, X_test_raw, y_train, y_test = train_test_split(X, y, train_size = 0.80)

print(f"- The training set consists of the {X_train_raw.shape[0]/features.shape[0]:.2f} of the initial dataset.")

- The training set consists of the 0.80 of the initial dataset.


## Ερώτημα 3

> *Πραγματοποιείστε προεπεξεργασία στα δεδομένα χρησιμοποιώντας τη μέθοδο TfidVectorizer της βιβλιοθήκης Scikit-Learn. Περισσότερες πληροφορίες για τη μέθοδο TfidVectorizer, που περιλαμβάνεται στις δηλώσεις του προγράμματος, μπορούν να βρεθούν [εδώ](#https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html). Να σημειωθεί πως η μέθοδος TfidVectorizer της Scikit-Learn αφαιρεί και τα σημεία στίξης, καθώς επεξεργάζεται το dataset.*

In [7]:
vectorizer = TfidfVectorizer()
X_train = vectorizer.fit_transform(X_train_raw)
X_test = vectorizer.transform(X_test_raw)

## Ερώτημα 4

> *Εκπαιδεύεστε το μοντέλο logistic regression.*

In [8]:
lr = LogisticRegression()
lr.fit(X_train, y_train)

LogisticRegression()

## Ερώτημα 5

> *Υπολογίστε την ακρίβειά του πάνω στο test set.*

In [9]:
lr.score(X_test, y_test)

0.968609865470852